Add clipboard support to //deform and //brush deform

TomyLobo opened the pull request:
This adds a -l flag to `//deform` and `//brush deform` which, when used, will fetch blocks from the clipboard instead of the world.
Additionally, coordinate transformations that would normally be based on the selection position+size are based on the clipboard's position+size when using that mode.

Example usage:
![image](https://user-images.githubusercontent.com/963797/223618869-af48bc3f-7d72-494a-a6be-e4d19094fbb8.png)
The shape on the right is the result of copying the square shape on the left and deforming the selection with the command on screen.

![image](https://user-images.githubusercontent.com/963797/223619577-038dd6fc-75d5-426c-b203-0dcfffc909d1.png)
This is the following brush using the same clipboard contents as before:
`//brush deform -l sphere 6 x=-x;y=0;z=-z`

me4502 commented on worldedit-core/src/main/java/com/sk89q/worldedit/function/factory/Deform.java:
```diff
@@ -64,6 +70,10 @@ public Deform(Extent destination, Region region, String expression) {
+        this(destination, region, expression, mode, false);
+    }
+
+    public Deform(Extent destination, Region region, String expression, Mode mode, boolean useClipboard) {
```
I sort of feel a sourceExtent & sourceRegion would be more useful here as something to pass in rather than useClipboard. Would also allow deform brushes to "bake in" the clipboard so you can have multiple different ones

TomyLobo replied:
Sorry, I only just saw your comment.
We'd have to offer both, because some people would want the current clipboard, some wouldn't.
Same with the -o flag.
That'd probably a separate PR.

wizjany replied:
baking in is the default for most things and should probably be consistent for ux. (see clipboard brush, `#sel` vs `#dsel` masks, etc)

TomyLobo replied:
ok I wasn't aware `//brush clipboard` already baked in the clipboard.
I use brushes so rarely :)
guess I'll have to bake that in before merging it then

TomyLobo replied:
`-o`, on the other hand, is not so simple.
1. Do we read placement type at `//brush` time or at click time?
2. Do we read `pos1` at `//brush` time or at click time?
3. Do we read the clipboard at `//brush` time or at click time?
4. Do we read `player` position at `//brush` time or at click time?

1. Probably at `//brush` time 
2. That'd be the difference between `#sel` and `#dsel`, so we'd probably want to offer both
3. We said we want to bake it in at `//brush` time, but if we offer both in 2., offering both here as well actually saves code
4. I have no idea tbh

TomyLobo commented:
rebased

TomyLobo commented:
rebased this on top of most of the //placement PR in order to be able to use the new Placement system for the Deform brush.
Also marked as draft because the Current code is a WIP

TomyLobo commented:
Moved the (contentious and not yet rebased) `//brush deform` changes into a separate `deform-brush-clipboard` branch and rebased again.
I'll unmark this as a draft once I'm done testing if it still works.

I might rebase `deform-brush-clipboard` later, but that's somewhat of a failed implementation and probably should be redone from scratch anyway.

me4502 commented on worldedit-core/src/main/java/com/sk89q/worldedit/math/transform/SimpleTransform.java:
```diff
@@ -0,0 +1,75 @@
+/**
+ * A more light-weight {@link Transform} than {@link AffineTransform}, supporting only translation and scaling.
+ */
+public class SimpleTransform implements Transform {
```
Is SimpleTransform able to please be a record? it might allow the JVM to optimise `isIdentity` to a computed constant

TomyLobo replied:
actually already had this turned into a record locally :)

TomyLobo commented:
Checks are passing, deform-brush-clipboard pushed
deform-brush-clipboard is untested.
deform-clipboard (this PR) is tested and working well, except for one issue:
for flat clipboards (i.e. clipboards which are only 1 thick on any axis) non-flat selections map some points outside the clipboard.
A workaround for the user is to set the flat axis to 0.

Two alternatives were considered:
option B: adjust only when inverting
advantage: `//deform -l 1` fills the entire selection even if the clipboard only is 1 thick
disadvantage(?): any coordinate on the axis in which the clipboard only is 1 thick will yield the same raw coordinate, meaning it's impossible to go outside the box on that axis anymore, with or without `-l`
disadvantage: the transform's implied matrix is singular if the clipboard only is 1 thick
disadvantage: the resulting transform is technically no longer really its own inverse (though `transform.inverse().inverse() == transform` still holds true)

option C: add 0.5 to the region/clipboard's min/max
advantage: no singularities, continuous, no special cases
disadvantage: breaks people's existing expressions for `//generate`,  `//generatebiome`, `//deform` and `//brush deform`
disadvantage: makes the relationship between block coordinates and -1..1 coordinates more complex
disadvantage: coordinates never reach -1 or 1, breaking expressions that rely on that

me4502 commented on worldedit-core/src/main/java/com/sk89q/worldedit/EditSession.java:
```diff
@@ -2419,31 +2500,47 @@ public int deformRegion(final Region region, final Vector3 zero, final Vector3 u
-            final Vector3 scaled = position.toVector3().subtract(zero).divide(unit);
-
             // transform
+            final Vector3 scaled = outputTransformInverse.apply(position.toVector3());
```
is `scaled` the correct name for this field now, if it's also translated? would `transformedPosition` be a more descriptive name?

TomyLobo replied:
It's called `inputPosition` now and after the expression is evaluated, I introduced an intermediate variable called `outputPosition` to match it.
Resolving conversation

me4502 commented on worldedit-core/src/main/java/com/sk89q/worldedit/EditSession.java:
```diff
@@ -2419,31 +2500,47 @@ public int deformRegion(final Region region, final Vector3 zero, final Vector3 u
 
-            // read block from world
-            final BaseBlock material = world.getFullBlock(sourcePosition);
+            // read block from world/clipboard
```
this should probably just say extent, given there could theoretically be other extents

TomyLobo replied:
it says "read block from source extent (e.g. world/clipboard)" now.
Resolving conversation

TomyLobo commented:
Now that #2582 is merged, I rebased and force-pushed this and #2580
Will test now, but I don't expect any issues, unless the heat is getting to me more than I thought :)

TomyLobo commented:
I'm done testing and enabled auto-merge.
So once someone reviews this, it's in :)

me4502 reviewed (commented):
This looks good to me, but I'd prefer if someone else reviewed this too as I'm less familiar with the expression/deform system

TomyLobo commented:
just [force-pushed](https://github.com/EngineHub/WorldEdit/compare/ef8b77b0f3ab470a8981a0ebde525744fef7ebfa..11a442591cd4b04f68bdb7db466c780e0b11cd9d) to change a javadoc description from input/output to source/target

TomyLobo commented:
[force-pushed](https://github.com/EngineHub/WorldEdit/compare/11a442591cd4b04f68bdb7db466c780e0b11cd9d..2e0ecdb02be70ecaa0911aca1901409276783d7e) to rename SimpleTransform to ScaleAndTranslateTransform, as suggested by @wizjany.

TomyLobo commented:
Any objections to merging this?

octylFractal commented:
I'd like to give it a review, but I can't immediately because I'm working.

TomyLobo commented:
this is still done and just pending a review :)

TomyLobo commented:
Rebased

TomyLobo commented on worldedit-core/src/main/java/com/sk89q/worldedit/regions/shape/WorldEditExpressionEnvironment.java:
```diff
@@ -23,24 +23,28 @@
     private Vector3 current = Vector3.ZERO;
     private final Extent extent;
 
+    @Deprecated
```
Done, as well as all the other @Deprecated annotations I added

octylFractal replied:
This links to the wrong method now?

TomyLobo commented on worldedit-core/src/main/java/com/sk89q/worldedit/EditSession.java:
```diff
@@ -2267,6 +2270,7 @@ public List<Countable<BlockState>> getBlockDistribution(Region region, boolean s
      * @throws ExpressionException if there is a problem with the expression
      * @throws MaxChangedBlocksException if the maximum block change limit is exceeded
      */
+    @Deprecated
```
Adjusted all @deprecated javadocs that link to an overload with a Transform and that are on methods that take zero/unit

octylFractal commented on worldedit-core/src/main/java/com/sk89q/worldedit/regions/shape/WorldEditExpressionEnvironment.java:
```diff
@@ -23,24 +23,31 @@
     private final Extent extent;
 
+    /**
+     * @deprecated Use {@link EditSession#makeBiomeShape(Region, Transform, BiomeType, String, boolean, int)} and pass a {@link ScaleAndTranslateTransform}.
```
Not confident that resolves properly, please check the output.

me4502 replied:
I can confirm this resolves properly- should be good to commit as-is
